prow/gcsupload/run.go

```diff
@@ -0,0 +1,145 @@
+	case kube.PathStrategyLegacy:
+		builder = gcs.NewLegacyRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	case kube.PathStrategySingle:
+		builder = gcs.NewSingleDefaultRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	}
+
+	spec, err := downwardapi.ResolveSpecFromEnv() // TODO: pass in all the config instead of needing this?
+	if err != nil {
+		return fmt.Errorf("could not resolve job spec: %v", err)
+	}
+
+	var gcsPath string
+	jobBasePath := gcs.PathForSpec(spec, builder)
+	if o.SubDir == "" {
+		gcsPath = jobBasePath
+	} else {
+		gcsPath = path.Join(jobBasePath, o.SubDir)
+	}
+
+	uploadTargets := map[string]gcs.UploadFunc{}
+
+	// ensure that an alias exists for any
+	// job we're uploading artifacts for
+	if alias := gcs.AliasForSpec(spec); alias != "" {
+		fullBasePath := "gs://" + path.Join(o.Bucket, jobBasePath)
+		uploadTargets[alias] = gcs.DataUpload(strings.NewReader(fullBasePath))
+	}
+
+	if latestBuilds := gcs.LatestBuildForSpec(spec, builder); len(latestBuilds) > 0 {
+		for _, latestBuild := range latestBuilds {
+			uploadTargets[latestBuild] = gcs.DataUpload(strings.NewReader(spec.BuildId))
+		}
+	}
+
+	for _, item := range o.Items {
+		info, err := os.Stat(item)
+		if err != nil {
+			logrus.Warnf("Encountered error in resolving items to upload for %s: %v", item, err)
+			continue
+		}
+		if info.IsDir() {
+			gatherArtifacts(item, gcsPath, info.Name(), uploadTargets)
+		} else {
+			destination := path.Join(gcsPath, info.Name())
+			if _, exists := uploadTargets[destination]; exists {
+				logrus.Warnf("Encountered duplicate upload of %s, skipping...", destination)
+				continue
+			}
+			uploadTargets[destination] = gcs.FileUpload(item)
+		}
+	}
+
+	for destination, upload := range extra {
+		uploadTargets[path.Join(gcsPath, destination)] = upload
+	}
+
+	if !o.DryRun {
+		ctx := context.Background()
+		gcsClient, err := storage.NewClient(ctx, option.WithCredentialsFile(o.GcsCredentialsFile))
+		if err != nil {
+			return fmt.Errorf("could not connect to GCS: %v", err)
+		}
+
+		if err := gcs.Upload(gcsClient.Bucket(o.Bucket), uploadTargets); err != nil {
+			return fmt.Errorf("failed to upload to GCS: %v", err)
+		}
+	} else {
+		for destination := range uploadTargets {
+			logrus.WithField("dest", destination).Info("Would upload")
+		}
+	}
+
+	logrus.Info("Finished upload to GCS")
+	return nil
+}
+
+func gatherArtifacts(artifactDir, gcsPath, subDir string, uploadTargets map[string]gcs.UploadFunc) {
+	logrus.Printf("Gathering artifacts from artifact directory: %s", artifactDir)
+	filepath.Walk(artifactDir, func(fspath string, info os.FileInfo, err error) error {
+		if info == nil || info.IsDir() {
+			return nil
+		}
+
+		// we know path will be below artifactDir, but we can't
+		// communicate that to the filepath module. We can ignore
+		// this error as we can be certain it won't occur and best-
+		// effort upload is OK in any case
+		if relPath, err := filepath.Rel(artifactDir, fspath); err == nil {
```

Comment:
Consider inverting this to keep indentation low:
```
relPath, err := filepath.Rel()
if err != nil { logrus.Warn(); return nil }
destination := path.Join()
...
```

Reply:
Can tackle in follow-up.